### PR TITLE
Feature/threaded

### PR DESCRIPTION
**Pull request by unapproachable**
Initial implementation of Threaded file operations to improve performance for directories with a large number of files.  This implementation is specifically and consciously only threading the file operations _per directory_.  This is to avoid any concurrency or race conditions and destination collisions.  The code should be fairly straightforward and uses the basic implementation of [ThreadPoolExecutor](https://docs.python.org/3/library/concurrent.futures.html).  This implementation is intended for improvements on a single-core and does not intend to leverage more CPU cores for improved I/O throughput.

In order to test and validate the speed up, this PR also included additional operation tracking (counting) and reporting.  It expanded the scope of this PR, but instrumentation was necessary to report the number of files seen, duplicated, and moved/copied in addition to a _throughput_ estimation for comparison of various thread pool sizes per run (see below).

Debugging can now be enabled via environment variables in addition to using switches.  This allows for scripts to be executed unmodified while toggling various levels of logging in the shell.

#### Linux
```shell
export LOGLEVEL=DEBUG
export LOGLEVEL=
```

#### Windows
```shell
set LOGLEVEL=DEBUG
set LOGLEVEL=
```

Lastly, version logging was add to the debug logging for easier traceability between developers/releases.

```console
[2021-08-31 12:13:22] - [DEBUG] - Debug logging output enabled.
[2021-08-31 12:13:22] - [DEBUG] - Running Phockup version 1.6.5
```

### Usage: 

```console
-c 1-255, --max-concurrency 1-255  Sets the level of concurrency for processing files in a directory.  Defaults to 1.  
Higher values can improve throughput of file operations
```

This PR sets the default to 1 thread, which should mimic the existing behavior (with slight overhead for including threading).  Useful settings need to be determined by the end user, but significant and measurable improvements can be achieved with as few as 5 threads.  Performance is affected by factors such as disk I/O speed (SSD vs HDD), network transfer rate, etc.  The more I/O wait, the more improvement I've seen using more threads.

Using the my own test data, these three runs were executed doing a dry run copying from a local SSD to a Samba network drive in order to test 1, 5, and 10 threads:

**Default:**
```shell
phockup.py C:\Pictures Z:\ -o -m -y 
```
<pre>
[2021-08-31 12:13:22] - [DEBUG] - Debug logging output enabled.
[2021-08-31 12:13:22] - [DEBUG] - Running Phockup version 1.6.5
[2021-08-31 12:13:22] - [WARNING] - Dry-run phockup (does a trial run with no permanent changes)...
<i>&lt;snip&gt;</i>
[2021-08-31 12:13:39] - [DEBUG] - Processed 43 files in <b>16.93 seconds</b>. Average Throughput: <b>2.54 files/second</b>
[2021-08-31 12:13:39] - [DEBUG] - Found 8 duplicate files.
[2021-08-31 12:13:39] - [DEBUG] - Would have moved 35 files.
</pre>

**5 Threads:**
```shell
phockup.py C:\Pictures Z:\ -o -m -y  --max-concurrency=5
```
<pre>
[2021-08-31 12:14:39] - [DEBUG] - Debug logging output enabled.
[2021-08-31 12:14:39] - [DEBUG] - Running Phockup version 1.6.5
[2021-08-31 12:14:39] - [INFO] - Using <b>5 workers</b> to process files.
[2021-08-31 12:14:39] - [WARNING] - Dry-run phockup (does a trial run with no permanent changes)...
<i>&lt;snip&gt;</i>
[2021-08-31 12:14:45] - [DEBUG] - Processed 43 files in <b>6.01 seconds</b>. Average Throughput: <b>7.16 files/second</b>
[2021-08-31 12:14:45] - [DEBUG] - Found 8 duplicate files.
[2021-08-31 12:14:45] - [DEBUG] - Would have moved 35 files.
</pre>

**10 Threads:**
```shell
phockup.py C:\Pictures Z:\ -o -m -y  --max-concurrency=10
```
<pre>
[2021-08-31 12:16:01] - [DEBUG] - Debug logging output enabled.
[2021-08-31 12:16:01] - [DEBUG] - Running Phockup version 1.6.5
[2021-08-31 12:16:01] - [INFO] - Using <b>10 workers</b> to process files.
[2021-08-31 12:16:01] - [WARNING] - Dry-run phockup (does a trial run with no permanent changes)...
<i>&lt;snip&gt;</i>
[2021-08-31 12:16:07] - [DEBUG] - Processed 43 files in <b>6.44 seconds</b>. Average Throughput: <b>6.68 files/second</b>
[2021-08-31 12:16:07] - [DEBUG] - Found 8 duplicate files.
[2021-08-31 12:16:07] - [DEBUG] - Would have moved 35 files.
</pre>

Three unit tests were added to help validate the functionality:

```python
def test_maxdepth_one():
    shutil.rmtree('output', ignore_errors=True)
    Phockup('input', 'output', maxdepth=1)
    validate_copy_operation()
    shutil.rmtree('output', ignore_errors=True)


def test_maxconcurrency_zero():
    shutil.rmtree('output', ignore_errors=True)
    with pytest.raises(ValueError, match="max_workers must be greater than 0"):
        Phockup('input', 'output', max_concurrency=0)


def test_maxconcurrency_five():
    shutil.rmtree('output', ignore_errors=True)
    Phockup('input', 'output', max_concurrency=5)
    validate_copy_operation()
    shutil.rmtree('output', ignore_errors=True)
```

Please let me know what I can do to help get this integrated into the main branch.

### CODE REVIEW

**Comment by ivandokov** [comment]
That's really interesting PR and I thank you for it 👏🏻!
I'll need more people to give their opinions as this looks like a big change and I don't want my decision to be the deciding factor here as I am not that experienced with the language.

**Comment by roykrikke** [comment]
Hello @unapproachable, as @ivandokov wrote this comment yesterday, this is an interesting PR 👏🏻! To be honest, I was already thinking about how we could add Multithreading support. I made it way too complex in my head and I think your solution is a nice pragmatic solution.

I've had a good look at your code and have a few questions or comments. 

> **First question:**

```
# Legacy string formatting syntax
# pytest-mock v3.6.1  that was introduced recently.  When tests
# are run, it throww the following error:
# TypeError: unsupported format string passed to MagicMock.__format__
# Revert to preferred syntax once MagicMock is updated
# Preferred syntax
```
Is listed in you code base. I'm also running with `pytest-mock v3.6.1` and I don't experience the same issue as you describe (see below). Am I missing something?

```
roy@computron:~/unapproachable/phockup$ pytest --cov-report term-missing:skip-covered --cov=src tests/
=============================================== test session starts ===============================================
platform linux -- Python 3.7.3, pytest-6.2.5, py-1.10.0, pluggy-1.0.0
rootdir: /home/roy/unapproachable/phockup
plugins: mock-3.6.1, cov-2.12.1
collected 52 items

tests/test_date.py ................                                                                         [ 30%]
tests/test_exif.py ....                                                                                     [ 38%]
tests/test_phockup.py ................................                                                                     [100%]

----------- coverage: platform linux, python 3.7.3-final-0 -----------
Name             Stmts   Miss  Cover   Missing
----------------------------------------------
src/date.py         75      3    96%   112, 115-116
src/exif.py         17      1    94%   14
src/phockup.py     198      7    96%   88-92, 134, 227-230
----------------------------------------------
TOTAL              296     11    96%

2 files skipped due to complete coverage.


======================================================= 52 passed in 8.88s =======================================================
roy@computron:~/unapproachable/phockup$
```
> **Second question:**

If we press `Ctrl-C` in an active run it is not stopping the prints but keeps printing lines and then stops. Which makes sense, I also had this problem with https://github.com/ivandokov/phockup/pull/118

With code:

```
 with concurrent.futures.ThreadPoolExecutor(max_workers=self.max_concurrency) as executor:
                executor.map(self.process_file, file_paths_to_process)
```
As far as I can judge. Is there always one extra process started on top of the main one. Therefore, the current implementation of  `Ctrl-C` does not work properly.
It would be nice if we could find a solution for this together. Do you have ideas for this?

> **First comment:**

We should not forget to update the readme.md with this feature.

> **Second comment:**

It would be really cool if we could merge this feature with the https://github.com/ivandokov/phockup/pull/118 (which is now part of the master branch). I committed. Did you already see this new feature? 

> **Third comment:**
In function `def get_file_type(self, mimetype):` Did you choose a different naming convention than in the rest of the file. That doesn't reads nice.
Also, I don't really understand why you changed the implementation. Can you explain the reason?


**All of the above is written in enthusiasm and in no way intended as criticism. On the contrary, I'm enthusiastic about your committed feature.**

**Comment by unapproachable** [comment]
@roykrikke My apologies, as I missed this update when you posted it and have only recently been working on it.  Let me see if I can address the points raised:

1. Running the code again, I also don't get the exception so maybe some sub-dependency has been updated?  After everything else is resolved, I'll put the fstring syntax back in and verify that it works.  This seems to be a common error with various `__format__` calls based on my perusal of StackOverflow, but maybe I was passing in data that made it particularly unhappy.
2. I haven't tested the CTRL-C case, but I will.  With this threading approach, we don't actually spawn more processes, so there should be an opportunity to cancel the execution with some sort of _magic_™.  The other options for parallelization certainly cause more issues tracking down the workers, but this should be solvable when using a threadpool.  I'll report back.
3. I actually have the readme updated, but I didn't commit it yet based on feedback.  Once we get this resolved, I'll include that in the PR.  I had initially forgotten and then I realized that I didn't have dominion over the flag names or even the processing model, so I waited to finalize that documentation.  I will included it in my update to the PR.
4. I'll need input on this one, as I did work on this the past couple of days and it's non-trivial.  Firstly, do we agree that we want a progress bar per-thread in the display?  So, `-c 10` would yield (up to) 10 progress lines in the console?  The second challenge I'm having is that we're passing the progress bar around to the `process_file` method.  My code didn't consider having 'state' passed into the worker, and recreating the pbar for every directory walk seems expensive.  I'll do my best, but I will want feedback on whatever compromises are necessary to make a multithreaded feature update a serially updated progress bar.
5. I did not implement or work on `get_file_type`.  Annotations suggest that was 'pabera'.

Unrelated to the above, I'd appreciate anyone other than myself to do some benchmarks/tests with this version of code to measure the benefit you see by adding additional workers.  I want to make sure others see the improvements that I see.  I was comparing 1, 5, and 10 workers to gauge scaling on a fixed workload.

Also unrelated, it appears that [test_phockup.py](https://github.com/ivandokov/phockup/blob/master/tests/test_phockup.py) has two declarations of `test_progress`, the first of which fails to run successfully.

Lastly, I absolutely appreciate the feedback and enjoy the collaboration.  I am *not* a Python developer, so this is strictly for mental exercise while working to manage my collection of images.  Let me convert this PR to a draft and make some changes to try an accommodate the asks and solve some of the identified issues (if possible).  And again, apologies for the delayed response.

**Comment by roykrikke** [comment]
> @roykrikke My apologies, as I missed this update when you posted it and have only recently been working on it. Let me see if I can address the points raised:

No problem at all! It's nice to work together on this super handy project, to make it even better :smile:


>     1. Running the code again, I also don't get the exception so maybe some sub-dependency has been updated?  After everything else is resolved, I'll put the fstring syntax back in and verify that it works.  This seems to be a common error with various `__format__` calls based on my perusal of StackOverflow, but maybe I was passing in data that made it particularly unhappy.

I haven't looked into this further. I rather complete this PR with you.

>     2. I haven't tested the CTRL-C case, but I will.  With this threading approach, we don't actually spawn more processes, so there should be an opportunity to cancel the execution with some sort of _magic_™.  The other options for parallelization certainly cause more issues tracking down the workers, but this should be solvable when using a threadpool.  I'll report back.

I offer myself as a tester :blush:

>     3. I actually have the readme updated, but I didn't commit it yet based on feedback.  Once we get this resolved, I'll include that in the PR.  I had initially forgotten and then I realized that I didn't have dominion over the flag names or even the processing model, so I waited to finalize that documentation.  I will included it in my update to the PR.

Absolutely perfect.

>     4. I'll need input on this one, as I did work on this the past couple of days and it's non-trivial.  Firstly, do we agree that we want a progress bar per-thread in the display?  So, `-c 10` would yield (up to) 10 progress lines in the console?  The second challenge I'm having is that we're passing the progress bar around to the `process_file` method.  My code didn't consider having 'state' passed into the worker, and recreating the pbar for every directory walk seems expensive.  I'll do my best, but I will want feedback on whatever compromises are necessary to make a multithreaded feature update a serially updated progress bar.

I think one progress bar would be fine. I think several only cause confusion. In the end it's all about what has been processed. When I think about it from the user's point of view, I see no reason to have multiple bars.
Regarding the implementation. We can certainly make changes if we can add this feature. This feature is very valuable for phockup. In addition, we tested the unit to ensure that it continues to work as it should.

> Unrelated to the above, I'd appreciate anyone other than myself to do some benchmarks/tests with this version of code to measure the benefit you see by adding additional workers. I want to make sure others see the improvements that I see. I was comparing 1, 5, and 10 workers to gauge scaling on a fixed workload.

I saw more or less the same speed difference. That's why I'm so passionate about your PR.  😁 

> Also unrelated, it appears that [test_phockup.py](https://github.com/ivandokov/phockup/blob/master/tests/test_phockup.py) has two declarations of `test_progress`, the first of which fails to run successfully.

Yes, that was a mistake on my side. Oops... See https://github.com/ivandokov/phockup/pull/128 about how it's going to be resolved.

> Lastly, I absolutely appreciate the feedback and enjoy the collaboration. I am _not_ a Python developer, so this is strictly for mental exercise while working to manage my collection of images. Let me convert this PR to a draft and make some changes to try an accommodate the asks and solve some of the identified issues (if possible). And again, apologies for the delayed response.

More or less the same on my side. In addition, I find phockupto be a really useful tool.

**Comment by roykrikke** [comment]
@unapproachable  Forgot to share the link below. Think this is valuable.
https://stackoverflow.com/questions/51601756/use-tqdm-with-concurrent-futures

**Comment by ivandokov** [comment]
I also think a single progress bar  is more user-friendly.

**Comment by roykrikke** [comment]
Hello @unapproachable I'm testing the new code. Looks really good. The only thing is the trigger Ctrl-C interrupt is not working for the printing on the cli. The idea works perfectly. Only printing is not what we want.

```
roy@computron:~/unapproachable/phockup$ ./phockup.py --dry-run --move 2017/scan/ /01_photos_and_videos/
[2021-09-15 20:21:24] - [WARNING] - Dry-run phockup (does a trial run with no permanent changes)...
[2021-09-15 20:21:25] - [INFO] - /selectie/DSC04200.JPG => skipped, duplicated file /01_photos_and_videos/2017/03/25/20170325-120427.jpg
[2021-09-15 20:21:25] - [INFO] - /selectie/DSC04214.JPG => skipped, duplicated file /01_photos_and_videos/2017/03/25/20170325-120802.jpg
[2021-09-15 20:21:25] - [INFO] - /selectie/DSC04273.JPG => skipped, duplicated file /01_photos_and_videos/2017/03/25/20170325-121912.jpg
[2021-09-15 20:21:26] - [INFO] - /selectie/DSC04277.JPG => skipped, duplicated file /01_photos_and_videos/2017/03/25/20170325-121920.jpg
[2021-09-15 20:21:26] - [INFO] - /selectie/DSC04307.JPG => skipped, duplicated file /01_photos_and_videos/2017/03/25/20170325-133811-2.jpg
[2021-09-15 20:21:26] - [INFO] - /selectie/DSC04327.JPG => skipped, duplicated file /01_photos_and_videos/2017/03/25/20170325-133901.jpg
^C[2021-09-15 20:21:27] - [INFO] - Received interupt. Shutting down...
[2021-09-15 20:21:27] - [INFO] - Processed 7 files in 2.12 seconds. Average Throughput: 3.31 files/second
[2021-09-15 20:21:27] - [INFO] - Found 6 duplicate files.
[2021-09-15 20:21:27] - [INFO] - selectie/DSC04335.JPG => skipped, duplicated file 01_photos_and_videos/2017/03/25/20170325-133913.jpg
roy@computron:~/unapproachable/phockup$
```
The last line needs to be before` [2021-09-15 20:21:27] - [INFO] - Received interupt. Shutting down...`

If I add `-c 2`
```
roy@computron:~/unapproachable/phockup$ ./phockup.py --dry-run --move -c 2 /2017/scan/ /01_photos_and_videos/
[2021-09-15 20:24:26] - [INFO] - Using 2 workers to process files.
[2021-09-15 20:24:26] - [WARNING] - Dry-run phockup (does a trial run with no permanent changes)...
[2021-09-15 20:24:26] - [INFO] - /scan/25270325/selectie/DSC04214.JPG => skipped, duplicated file /01_photos_and_videos/2017/03/25/20170325-120802.jpg
[2021-09-15 20:24:26] - [INFO] - /scan/25270325/selectie/DSC04200.JPG => skipped, duplicated file /01_photos_and_videos/2017/03/25/20170325-120427.jpg
[2021-09-15 20:24:27] - [INFO] - /scan/25270325/selectie/DSC04273.JPG => skipped, duplicated file /01_photos_and_videos/2017/03/25/20170325-121912.jpg
[2021-09-15 20:24:27] - [INFO] - /scan/25270325/selectie/DSC04277.JPG => skipped, duplicated file /01_photos_and_videos/2017/03/25/20170325-121920.jpg
[2021-09-15 20:24:27] - [INFO] - /scan/25270325/selectie/DSC04327.JPG => skipped, duplicated file /01_photos_and_videos/2017/03/25/20170325-133901.jpg
[2021-09-15 20:24:27] - [INFO] - /scan/25270325/selectie/DSC04307.JPG => skipped, duplicated file /01_photos_and_videos/2017/03/25/20170325-133811-2.jpg
[2021-09-15 20:24:27] - [INFO] - /scan/25270325/selectie/DSC04335.JPG => skipped, duplicated file /01_photos_and_videos/2017/03/25/20170325-133913.jpg
[2021-09-15 20:24:28] - [INFO] - /scan/geen datum/P1100308.JPG => /01_photos_and_videos/unknown/p1100308.jpg
[2021-09-15 20:24:28] - [INFO] - /scan/geen datum/P1100307.JPG => /01_photos_and_videos/unknown/p1100307.jpg
^Cpanic: fold_constants JMPENV_PUSH returned 2.
[2021-09-15 20:24:28] - [INFO] - Received interupt. Shutting down 2 workers...
[2021-09-15 20:24:28] - [INFO] - Processed 42 files in 1.63 seconds. Average Throughput: 25.78 files/second
[2021-09-15 20:24:28] - [INFO] - Found 7 duplicate files.
[2021-09-15 20:24:28] - [INFO] - Would have moved 2 files.
[2021-09-15 20:24:28] - [INFO] - /2017/scan/geen datum/P1100310.JPG => /01_photos_and_videos/unknown/P1100310.JPG
[2021-09-15 20:24:28] - [INFO] - /2017/scan/geen datum/P1100309.JPG => /01_photos_and_videos/unknown/P1100309.JPG
```


If I add `--progress` the same thing.
```
roy@computron:~/unapproachable/phockup$ ./phockup.py --dry-run --move --progress /2017/scan/ /01_photos_and_videos/
[2021-09-15 20:26:05] - [WARNING] - Dry-run phockup (does a trial run with no permanent changes)...
/2017/scan/25270325/selectie/DSC04200.JPG => skipped, duplicated file /01_photos_and_videos/2017/03/25/20170325-120427.jpg
/2017/scan/25270325/selectie/DSC04214.JPG => skipped, duplicated file /01_photos_and_videos/2017/03/25/20170325-120802.jpg
/2017/scan/25270325/selectie/DSC04273.JPG => skipped, duplicated file /01_photos_and_videos/2017/03/25/20170325-121912.jpg
/2017/scan/25270325/selectie/DSC04277.JPG => skipped, duplicated file /01_photos_and_videos/2017/03/25/20170325-121920.jpg
Progressing: '/media/royandannelies/100_UITZOEKEN/Fotos/uitzoeken/2017/scan' :   0%|▏                                                  | 5/1544 [00:01<07:25,  3.45file/s]Progressing: '/media/royandannelies/100_UITZOEKEN/Fotos/uitzoeken/2017/scan' :   0%|▏                                                  | 5/1544 [00:01<07:41,  3.33file/s]
/2017/scan/25270325/selectie/DSC04307.JPG => skipped, duplicated file /01_photos_and_videos/2017/03/25/20170325-133811-2.jpg
roy@computron:~/unapproachable/phockup$
```

I know it it's an initial attempt to trigger Ctrl-C interrupt to threaded tasks. I think it's super cool that we're here. Nice clean implementation too. I will also think about how we can solve this trigger Ctrl-C interrupt and have the status printing also nice.

**Comment by unapproachable** [comment]
So, a couple thing about the CTRL-C thing we should discuss.  

Firstly, it's requiring unholy things to actually purge the task queue (see the `def handle_interrupt`).  Okay, maybe _just_ accessing private variables to purge the queues because no one thought to expose them publicly in the case we cancel the tasks). 
 I'm assuming we're okay with that since we're playing in the realm of SIGINT, etc.
 
Secondly, I haven't been able to *stop* executing of threads that are in-flight.  This is why we still get output after the CTRL-C is caught.  No new tasks are executed, but the ones that were scheduled still run.  This means we can have as many lines reported AFTER the interrupt is caught.

Options:

1. I can try more unholy manipulation of the ThreadPoolExecutor, its queues, etc.  I haven't looked the code much, but I may be able to purge the queues and "wait" for the threads to finish before logging messaging about it being "done".  I'm not sure how well this will work, but may be the most concise way to tackle the output issue.

2. I can _try_ to 'spoof' or buffer the output to make it _look_ correct by holding the interrupt messaging until we "know" the threads are done.  This would be synthetic and probably require some state flags, but still not sure I can make it do what I want.  Specifically, the messaging about throughput in your example (`print_action_report()`) comes after `walk_directories()` and we still get output from `process_file()` after that.  That is well out of execution flow, and far from being in scope to check state of the thread pool.  More hacking required and possibly more instance scope state trackers.  Seems horrible.  I regret suggesting it.

CTRL-C aside, is the progress bar working for you?  I'm on Windows, so I have to rely on you guys to test on Linux.  I had to resort to ASCII printing on this platform since the console doesn't support Unicode characters well for the animation.  Please do let me know how it's looking. 

Related to Windows, I need someone to the verify that the test harness passes on a non-Windows platform.  Symlinks are weird and I don't have the concept of "root" to test for so there is a bit of divergence in the test path.

Lastly, I wasn't sure if I should dump the statistics at the end of a run or not, but I added it in for my own testing purposes.  The progress bar has a nice display that is relevant (which is why I have `leave=True`), but without that, I thought it would be nice to print summary stats.

I'll work on getting CTRL-C to behave and testing with the latest commits.  Let me know if you spot anything else misbehaving.

**Comment by roykrikke** [comment]
Hello @unapproachable,

I would like to share on option three. You first option is not clean, but understand your logic. The second option is a lot of work. That is why I propose a third option. What if we want to stop (Ctrl-C) it but we give the application time to finish the last processes neatly and only then close everything. I had looked around on the internet and found the following link very instructive. That's why I'm sharing this one. https://github.com/wbenny/python-graceful-shutdown

The progress bar is working perfectly on my linux machine. Also the test harness passes on my linux machine perfectly. I also agree with the printing de stats if the `--progress` option is not used. We could have a `--stats` options with can't be used in combination with -`-progress`. Just an idea.

So far I haven't found any other things that don't work as they should. It's a bit of a shame that we have to do so much work to get Ctrl-C to work properly.

**Comment by unapproachable** [comment]
@roykrikke, thank you for the article.  It is very robust solution to interrupting a multi-threaded process.  I didn't actually attempt to implement it because it added even more overhead to allowing multiple files to be processed than was already introduced with ThreadPoolExecutor, and I didn't want to overload the 'simple' solution with a heavy dependency on `asyncio` in addition to `concurrent` libraries.  If we go to a truly multhi-processing model (taking advantage of multiple cores, etc.) then we should certainly revisit this type of solution.

Having said that, this prompted me to revisit the behavior of "CTRL-C" again from the 'top down'.  I realized that I could just catch the CTRL-C and nicely ask the ThreadPoolExecutor to shut itself down, telling it that I'd wait for it to finish before continuing on.  This allows the output to flow in the correct serial order to the console and only adds a slight delay to the response to the user's action of hitting CTRL-C.  

I've noticed significantly different behavior between Windows and Linux on how it cleans up threads, so I'll rely on you to let me know what your experience is with this latest attempt to get concurrent processing and progress bars included.  If this solution is close enough, we can move on to refactoring the `process_file` function which is a bit hard to follow with the `while True` and the many `break` calls.  I only noticed because I was trying to correctly count processed files and update the progress bar the correct number of times.

Please let me know what your experience is with this latest attempt at an acceptable user experience.

**Comment by roykrikke** [comment]
@unapproachable, my feedback so far.

First of all:
`pytest --cov-report term-missing:skip-covered --cov=src tests/` Runs perfect. Code coverage has decreased slightly. But I understand that you want to get it working first and then do some refactoring. So don't pay much attention to this point. That will come later.

With: `./phockup.py  -c 8 --move  /input/ /output/ ` and pressing Ctrl-C. works fine normally. Which looks very neat. But sometimes I see in linux `panic: fold_constants JMPENV_PUSH returned 2.` I have never seen that particular error before. It seems to indicate a race condition related to ThreadPoolExecutor.

With: `./phockup.py --progress -c 8 --move  /input/ /output/` and pressing Ctrl-C. works fine. But sometimes 
I see some lines are printed and also the progress bar is printed several times. With the progress option enabled I also see sometimes the  `panic: fold_constants JMPENV_PUSH returned 2.`

**Comment by unapproachable** [comment]
Unfortunately, I can't control the messages being displayed during the interrupt.  I'm honestly thinking these are stderr messages being echoed by the underlying C binaries or another lower level component used for OS IO (if you google that message, it brings up a bunch of Perl errors... yeah, Perl).  So I'm not sure we'll be able to gracefully terminate whatever is being interrupted since we're already catching the exception and handling it (by waiting for it to complete) and the messages (I have seen many different types of errors, in including what appears to be hex garbage, displayed to the screen).

```shell
^C^Cpanic: fold_constants JMPENV_PUSH returned 2 at -e line 639.
```

```shell
^CTerminating on signal SIGINT(2)
```

```shell
Out of memory!
Use of uninitialized value  in exists at C:/Perl/lib/Config.pm line 70.
^Terminating on signal SIGINT(2)
C^C^C^C^C^C
```
**NOTE:** I do not have Perl installed at `C:/Perl/` or anywhere else, for that matter.

My only thought on that is to eat stderr... which seems dangerous, but also worth considering since we aren't generating any of these errors and we can't seem to catch them in the exception block (probably related to their panic nature and/or being an external library).  I'm open to suggestions on how to clean up the output, but I fear that it's all or nothing with stderr during the file operations.

**Comment by roykrikke** [comment]
@unapproachable 

I think I found a clue. exiftool is a perl tool. Since we are executing exiftool externally from python we stop the process while executing exittool. 
From: https://exiftool.org/ 
_The stand-alone Windows executable does not require Perl._

We may be able to find a way to stop the python processes only when exiftool is really done.

What do you think?

**Comment by unapproachable** [comment]
@roykrikke HA!  That's exactly what I tracked down!  So, I 'solved' the issue by swallowing the exiftool stderr.  It's not great, but I left it exposed for non-concurrent operations in case someone needs to 'debug' it.  We can also ALWAYS swallow it, and allow a flag to turn it on, but even if exiftool fails, I'm not sure what we can do about it.  It would be nice to allow the user to at least report the error they are seeing by enabling a mode and/or verifying that it happens in both single/concurrent modes.

Please test my latest commit.  I think we're really close and I'd **love** to get this PR out of my queue.  I have more things I want to include, but don't enjoy spinning so many plates at the same time.  I'm likely to drop them all.

I look forward to any testing results!

**Comment by roykrikke** [comment]
This weekend I will to do some testing. I'll keep you informed.

**Comment by roykrikke** [comment]
Hello @unapproachable,

 I have tested thoroughly. As you wrote above. Looks like we're almost there. But unfortunately I have found an important detail that doesn't work quite well yet. 

With: `./phockup.py  --move  /input/ /output/ ` and pressing Ctrl-C. Does this not work! It just keeps executing. I need to keep Ctrl-C pressed and then after a while it stops executing. I can see the following for every interrupt:
`^C[2021-10-03 17:19:00] - [WARNING] - Received interupt. Shutting down...`

With: `./phockup.py  --progress --move  /input/ /output/ ` and pressing Ctrl-C. Does this not work! It just keeps executing. I need to keep Ctrl-C pressed and then after a while it stops executing. I can see the following for every interrupt:
`^C[2021-10-03 17:19:00] - [WARNING] - Received interupt. Shutting down...`

With: `./phockup.py  -c 8 --move  /input/ /output/ ` and pressing Ctrl-C. works fine normally. Which looks very neat.

With: `./phockup.py --progress -c 8 --move  /input/ /output/` and pressing Ctrl-C. works fine. This will print an extra progress bar.    I think we'll just have to accept that for now.

I myself have made an attempt to find out why it does not work well with one process. I parked it for now because I think I'm overlooking it. 

Side note:
Lines: 91-94, 131, 136-137,  231-235, 256-259 of `src/phockup.py ` should actually be included in unit testing.

**Comment by unapproachable** [comment]
Updated code to correct this... but now I'm realizing that we sometimes have weird messages even in single threaded mode.  Not sure if it's safe to swallow all stderr from exiftool or not, regardless of concurrency.

**Comment by roykrikke** [comment]
@unapproachable any news on this nice feature?

**Comment by unapproachable** [comment]
@roykrikke I'm going to have to rely on the community to finish up the work on coverage testing.  It's largely platform specific code, so the coverage will vary based on platform and I'm not well versed enough to manage the various test cases across multiplatform coverage testing.  Having said that, I feel that this feature was pretty much completed, pending community review.  I agree that it's very helpful and speeds things up tremendously.  Maybe we can put in a follow on issue, post-merge, to tackle passing 100% coverage?

It appears to still be a clean merge, but if there are issues pulling it in I'd be happy to sort them out once identified.  I'm happy to help support this feature post-merge and field feature requests, but this initial implementation is probably ready to be closed, one way or another.

**Comment by ivandokov** [comment]
Thank you @unapproachable and @roykrikke for all of the work related to this PR! The code introduced here is kind of out of my league but trust your input and I'm happy to merge the new feature! We can totally improve the tests post merge.

**Comment by iFloris** [comment]
Nice, this is a really big feature! My thanks and appreciation to you all! 👏

**Comment by SebastianChudoba** [comment]
I did some testing with --max-concurrency and experienced Errno 17 with any thread option larger than 2.
My system is attached.
Bildschirmfoto 2022-01-17 um 18 14 09

**Comment by SebastianChudoba** [comment]
I did some testing with --max-concurrency and experienced Errno 17 with any thread option larger than 2.
My system is attached.
Bildschirmfoto 2022-01-17 um 18 14 59

**Comment by roykrikke** [comment]
> I did some testing with --max-concurrency and experienced Errno 17 with any thread option larger than 1. Though, with the option set, the work is done faster than with the default. My system is attached. Bildschirmfoto 2022-01-17 um 18 14 59

This Errno?:
Description: Just checking if the directory already exist throws this error message [Errno 17] File exists because we are just checking if the directory name exist or not which will return the directory name of the mydir value being passed but not if it already exist or not. What is being missed is not checking if that directory already exist which can be done by checking the path with os.path.exists() and in there we passed the respective directory name. 

Could you share more details on you situation? Then maybe we can test together. I don't own an apple machine with maxOS.

**Comment by SebastianChudoba** [comment]
I was doing a --move operation while INPUTDIR and OUTPUTDIR were the same directory. /User/Photos/2022/...
It was simply to rename monthly folders from names to integers, i.e. JAN -> 01 and so on. I did this on the whole library while it was syncing to nextcloud. It had likely to do with having nextcloud deleting folders that were still called by phockup...
